### PR TITLE
MAPO presubmits on master run with CustomNoUpgrade

### PR DESCRIPTION
**Pull request by mdbooth**

### CODE REVIEW

**Comment by mdbooth** [comment]
```
function to_commas (
                   ^-- SC1036: '(' is invalid here. Did you forget to escape it?
```

Interestingly this shellcheck is incorrect. I used a subshell here deliberately as a cleaner way to avoid polluting the main shell with an alternate IFS. I'll change it to use {} and local instead, but meh.

**Comment by mdbooth** [comment]
```
                ^-------------^ SC2046: Quote this to prevent word splitting.
                            ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
```

Exploiting word splitting to change the whitespace into commas is the entire purpose of that function 🤦‍♂️ I think I'll have to disable these checks.

File: ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  as: openshift-e2e-openstack-customnoupgrade
+  steps:
+    pre:
+    - ref: ipi-conf-customnoupgrade
+    - chain: ipi-openstack-pre
+    test:
+    - chain: openstack-test-e2e
+    post:
+    - chain: ipi-openstack-post
+    env:
```
Comment (pierreprinetti) [inline ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml:11]: Did you mean to pass something as `CUSTOM_FEATURE_FLAGS_ENABLED`?

File: ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+workflow:
+  as: openshift-e2e-openstack-customnoupgrade
+  steps:
+    pre:
+    - ref: ipi-conf-customnoupgrade
```
Comment (pierreprinetti) [inline ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml:5]: What about having `ipi-conf-customnoupgrade` always in the `ipi-openstack-pre` chain, and having it turn into a no-op when both variables are `""`?

File: ci-operator/step-registry/ipi/conf/customnoupgrade/ipi-conf-customnoupgrade-ref.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  from: cli
+  resources:
+    requests:
+      cpu: 10m
+      memory: 100Mi
+  env:
+    - name: CUSTOM_FEATURE_FLAGS_ENABLED
+      default: ""
+    - name: CUSTOM_FEATURE_FLAGS_DISABLED
+      default: ""
```
Comment (pierreprinetti) [inline ci-operator/step-registry/ipi/conf/customnoupgrade/ipi-conf-customnoupgrade-ref.yaml:13]: Please add a `documentation` stanza specifying the required format. Which brings me to: Why don't you just directly accept a comma-separated list?

File: ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml (hunk 2 above)
Comment (mdbooth) [reply ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml:5]: That does feel more 'composed'.

File: ci-operator/step-registry/ipi/conf/customnoupgrade/ipi-conf-customnoupgrade-ref.yaml (hunk 3 above)
Comment (mdbooth) [reply ci-operator/step-registry/ipi/conf/customnoupgrade/ipi-conf-customnoupgrade-ref.yaml:13]: I worried about consistency so I did a quick look at other uses of lists in releases/. Here are the first 3 I looked at:

`AZURE_ARC_REGIONS`: space separated
`ADDON_PARAMETERS`: json string
`CLUSTERPOOL_LIST`: comma separated list

I stopped looking at this point. I conclude there is no consistency and I'm free to specify a comma separated list here.

File: ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml (hunk 2 above)
Comment (mdbooth) [reply ci-operator/step-registry/openshift/e2e/openstack/customnoupgrade/openshift-e2e-openstack-customnoupgrade-workflow.yaml:5]: That said, I think I recall getting pushback when I tried doing something similar before which resulted in me creating the ugly `-noramfs` ref. Lets see who objects.

**Comment by mdbooth** [comment]
The image substitution is definitely working in MAPO now: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/26574/rehearse-26574-pull-ci-openshift-machine-api-provider-openstack-master-e2e-openstack/1499007503445266432/artifacts/release/artifacts/release-payload-latest/0000_30_machine-api-operator_01_images.configmap.yaml

**Comment by mdbooth** [comment]
This appears to be working correctly. The machine controller deployment is using:

```json
"image": "registry.build01.ci.openshift.org/ci-op-vhy5nw0i/stable@sha256:51966c97d34b558a2c2f367f94e70fa4d5fd021bd7940b6201fad5551b7693f0",
```
See: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/26574/rehearse-26574-pull-ci-openshift-machine-api-provider-openstack-master-e2e-openstack/1499007503445266432/artifacts/e2e-openstack/gather-extra/artifacts/deployments.json

This is the deployed MAPO image:
```json
 "clusterAPIControllerMAPO": "registry.build01.ci.openshift.org/ci-op-vhy5nw0i/stable@sha256:51966c97d34b558a2c2f367f94e70fa4d5fd021bd7940b6201fad5551b7693f0",
```
See: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/26574/rehearse-26574-pull-ci-openshift-machine-api-provider-openstack-master-e2e-openstack/1499007503445266432/artifacts/release/artifacts/release-payload-latest/0000_30_machine-api-operator_01_images.configmap.yaml

This is the image we built:
```
2022-03-02T13:08:43.129144242Z Successfully pushed image-registry.openshift-image-registry.svc:5000/ci-op-vhy5nw0i/pipeline@sha256:51966c97d34b558a2c2f367f94e70fa4d5fd021bd7940b6201fad5551b7693f0
```
See: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/26574/rehearse-26574-pull-ci-openshift-machine-api-provider-openstack-master-e2e-openstack/1499007503445266432/artifacts/build-logs/openstack-machine-api-provider.log

**Comment by mdbooth** [comment]
I just pushed a change which changed the name of the manifest file to `manifest_customnoupgrade.yaml`, as I feel this is less likely to conflict with anything else in the future. There are no other changes.

**Comment by mdbooth** [comment]
Built image: `2022-03-02T16:26:18.628556997Z Successfully pushed image-registry.openshift-image-registry.svc:5000/ci-op-nx8lp864/pipeline@sha256:1a8a20407f841248ec4955f6ea0beb704c7efcf32f7bb66f0124c243be643fee
`

Running image: `                                "image": "registry.build01.ci.openshift.org/ci-op-nx8lp864/stable@sha256:1a8a20407f841248ec4955f6ea0beb704c7efcf32f7bb66f0124c243be643fee",
`

Looks good.

**Comment by pierreprinetti** [comment]
Great job! That will be useful, thank you.

/lgtm

**Comment by mdbooth** [comment]
/assign vrutkovs

Can you take a look at the ipi-conf-customnoupgrade step?

**Pull request merged**